util/rate_limiter.cc

```diff
@@ -46,34 +50,69 @@ struct GenericRateLimiter::Req {
   bool granted;
 };
 
+static std::unordered_map<std::string, OptionTypeInfo>
+    generic_rate_limiter_type_info = {
+#ifndef ROCKSDB_LITE
+        {"rate_bytes_per_sec",
+         {offsetof(struct GenericRateLimiter::GenericRateLimiterOptions,
+                   max_bytes_per_sec),
+          OptionType::kInt64T}},
+        {"refill_period_us",
+         {offsetof(struct GenericRateLimiter::GenericRateLimiterOptions,
+                   refill_period_us),
+          OptionType::kInt64T}},
+        {"fairness",
+         {offsetof(struct GenericRateLimiter::GenericRateLimiterOptions,
+                   fairness),
+          OptionType::kInt32T}},
+        {"auto_tuned",
+         {offsetof(struct GenericRateLimiter::GenericRateLimiterOptions,
+                   auto_tuned),
+          OptionType::kBoolean}},
+        {"clock",
+         OptionTypeInfo::AsCustomSharedPtr<SystemClock>(
+             offsetof(struct GenericRateLimiter::GenericRateLimiterOptions,
+                      clock),
+             OptionVerificationType::kByNameAllowFromNull,
+             OptionTypeFlags::kAllowNull)},
+#endif  // ROCKSDB_LITE
+};
+
 GenericRateLimiter::GenericRateLimiter(
     int64_t rate_bytes_per_sec, int64_t refill_period_us, int32_t fairness,
     RateLimiter::Mode mode, const std::shared_ptr<SystemClock>& clock,
     bool auto_tuned)
     : RateLimiter(mode),
-      refill_period_us_(refill_period_us),
-      rate_bytes_per_sec_(auto_tuned ? rate_bytes_per_sec / 2
-                                     : rate_bytes_per_sec),
-      refill_bytes_per_period_(
-          CalculateRefillBytesPerPeriod(rate_bytes_per_sec_)),
-      clock_(clock),
+      options_(rate_bytes_per_sec, refill_period_us, fairness, clock,
+               auto_tuned),
       stop_(false),
       exit_cv_(&request_mutex_),
       requests_to_wait_(0),
       available_bytes_(0),
-      next_refill_us_(NowMicrosMonotonic()),
-      fairness_(fairness > 100 ? 100 : fairness),
       rnd_((uint32_t)time(nullptr)),
       wait_until_refill_pending_(false),
-      auto_tuned_(auto_tuned),
       num_drains_(0),
-      prev_num_drains_(0),
-      max_bytes_per_sec_(rate_bytes_per_sec),
-      tuned_time_(NowMicrosMonotonic()) {
+      prev_num_drains_(0) {
+  RegisterOptions(&options_, &generic_rate_limiter_type_info);
   for (int i = Env::IO_LOW; i < Env::IO_TOTAL; ++i) {
     total_requests_[i] = 0;
     total_bytes_through_[i] = 0;
   }
+  Initialize();
+}
+void GenericRateLimiter::Initialize() {
+  if (options_.clock == nullptr) {
+    options_.clock = SystemClock::Default();
+  }
+  options_.fairness = std::min(options_.fairness, 100);
+  next_refill_us_ = NowMicrosMonotonic();
+  tuned_time_ = std::chrono::microseconds(NowMicrosMonotonic());
+  if (options_.auto_tuned) {
+    rate_bytes_per_sec_ = options_.max_bytes_per_sec / 2;
+  } else {
+    rate_bytes_per_sec_ = options_.max_bytes_per_sec;
```

Comment:
@mrambacher This can change the rate limit to an unexpected value. `SetBytesPerSecond()` modifies `rate_bytes_per_sec_` but doesn't modify `options_.max_bytes_per_sec`. Then, after an unrelated operation like persisting OPTIONS file, we can see `GenericRateLimiter::PrepareOptions()` is called. In turn it calls this, which resets `rate_bytes_per_sec_` to the rate limit originally provided in the constructor.

Reply:
Never mind, the `PrepareOptions()` is called on a different `GenericRateLimiter` object than the one in use by the DB. We still don't know how `rate_bytes_per_sec_` is getting reset to `options_.max_bytes_per_sec`.

Reply:
@ajkr Are you sharing a RateLimiter between DB instances?  I could see where that could potentially reset the rate (call PrepareOptions again).

It might also happen if you did a ConfigureDBOptionsFromMap/String against a RateLimiter where you had already called SetBytesPerSecond.

I looked at the DB::SetOptions/SetDBOptions code and do not see a path there.

Can you give any additional details on what may be happening when the value is reset?

Reply:
@mrambacher Thanks for the response. Yes it is shared across instances. I found out the user is calling `GetOptionsFromString()` to mutate an `Options` structure that  has the shared `rate_limiter` already configured, and the string they pass in does not involve `rate_limiter`. Here's a test case that shows the issue.

```
diff --git a/options/options_test.cc b/options/options_test.cc
index 0eeaca484..0f3c8f58d 100644
--- a/options/options_test.cc
+++ b/options/options_test.cc
@@ -4954,6 +4954,22 @@ TEST_F(ConfigOptionsTest, MergeOperatorFromString) {
   ASSERT_EQ(*delimiter, "&&");
 }
 
+TEST_F(ConfigOptionsTest, ConfiguringOptionsDoesNotRevertRateLimiterBandwidth) {
+  // Regression test for bug where rate limiter's dynamically set bandwidth
+  // could be silently reverted when configuring an options structure with an
+  // existing `rate_limiter`.
+  Options base_options;
+  base_options.rate_limiter.reset(
+      NewGenericRateLimiter(1 << 20 /* rate_bytes_per_sec */));
+  Options copy_options(base_options);
+
+  base_options.rate_limiter->SetBytesPerSecond(2 << 20);
+  ASSERT_EQ(2 << 20, base_options.rate_limiter->GetBytesPerSecond());
+
+  ASSERT_OK(GetOptionsFromString(base_options, "", &copy_options));
+  ASSERT_EQ(2 << 20, base_options.rate_limiter->GetBytesPerSecond());
+}
+
 INSTANTIATE_TEST_CASE_P(OptionsSanityCheckTest, OptionsSanityCheckTest,
                         ::testing::Bool());
 #endif  // !ROCKSDB_LITE
```